### PR TITLE
[WIP] Add mangle/demangle helper functions

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122796
* #123129
* #123110
* #123109
* #123108

- `mangle` mangles any string that matches `[A-z0-9.<>]+` into a valid
  Python identifier (e.g. has no special chars aside from underscore).
  We optimize for readability of the string in common cases. `demangle`
  is the inverse operation
- Why do we need <>? It's needed for nested functions.

In a future PR up in the stack, we use `mangle` to
generate a name that can be used an operator name. I split off the
mangling because that PR was getting long.

Test Plan:
- new tests

### CODE REVIEW

File: torch/_library/utils.py [hunk 1]
```diff
@@ -156,3 +157,67 @@ def zip_schema(schema, args, kwargs):
...
+
+
+def unique_name(fn: Callable) -> str:
+    """Given a callable, return a name that uniquely identifies it.
+
+    Some cases:
+    - a function is identified by its __module__ and __qualname__
+    - lambda requires the object's id to distinguish
+    """
+    is_lambda = fn.__name__ == "<lambda>"
```
Comment (albanD) [inline torch/_library/utils.py:217]: nit: check that the input is actually a callable?
Also is a Callable guaranteed to have a `__name__`/`__module__`/`__qualname__` attributes?

File: torch/_library/utils.py [hunk 2]
```diff
@@ -156,3 +157,67 @@ def zip_schema(schema, args, kwargs):
             continue
         yield info, args[i]
     return
+
+
+def mangle(value: str) -> str:
+    """Mangles a string into a usable Python identifier.
+
+    https://docs.python.org/3/reference/lexical_analysis.html#identifiers
```
Comment (albanD) [inline torch/_library/utils.py:165]: I'm surprised there is no generic tool to make any string into a indentifier...

File: torch/_library/utils.py [hunk 3]
```diff
@@ -156,3 +157,67 @@ def zip_schema(schema, args, kwargs):
...
+    value = value.replace("Z0", "<")
+    value = value.replace("Z1", ">")
+    value = value.replace("Z2", "@")
+    value = value.replace("X", ".")
+    value = value.replace("?", "X")
+    value = value.replace("!", "Z")
+    return value
+
+
+def unique_name(fn: Callable) -> str:
```
Comment (albanD) [inline torch/_library/utils.py:210]: callable_unique_name ?

File: test/test_custom_ops.py [hunk 4]
```diff
@@ -721,6 +722,40 @@ def _generate_examples(self, typ):
...
+            "torch.testing._internal.custom_op_db.get_fn2.<locals>.<lambda>@0xDEADBEEF",
+        ]
+
+        for value in examples:
+            mangled = utils.mangle(value)
+            self.assertTrue(mangled.isidentifier())
+            demangled = utils.demangle(mangled)
+            self.assertEqual(demangled, value)
+
+    def test_unique_name(self):
```
Comment (albanD) [inline test/test_custom_ops.py:743]: I would also test this with:
- method
- classmethod
- staticmethod
- builtin
- builtin method

**Comment by ezyang** [comment]
@zou3519 and I talked about this in our 1:1, and the conclusion was that we only really need mangling for top level identifiers, and a special case for one shot inner functions just for ease of testing.

File: torch/_library/utils.py (hunk 2 above)
Comment (Skylion007) [reply torch/_library/utils.py:165]: This doesn't handle if a value is a keyword, right? That can be detect with lib/keyword.py

File: torch/_library/utils.py [hunk 5]
```diff
@@ -156,3 +157,67 @@ def zip_schema(schema, args, kwargs):
...
+    - @ -> Z2
+
+    Examples:
+    - foo.bar.baz -> fooXbarXbaz
+    - foo_bar.baz -> foo_barXbaz
+    - X.Z -> ZXXZZ
+    - ZQ.ZQ -> ZZQXZZQ
+    - <lambda> -> Z0lambdaZ1
+    """
+    assert re.match(r"^[A-z0-9.<>@_]+$", value), "invalid value"
```
Comment (Skylion007) [inline torch/_library/utils.py:187]: Did you intend to include ```ASCII values from 91 to 96 ([, \, ], ^, _, and `)``` Or should this be replaced too?

File: torch/_library/utils.py (hunk 2 above)
Comment (Skylion007) [reply torch/_library/utils.py:165]: There is does github gist that seems to cover some more edgecases https://gist.github.com/JamesPHoughton/3a3f87c6662bf5c9eccc9f2206e228fd

**Comment by zou3519** [comment]
I am planning on turning this PR into the "automatic namespace support for custom_op" PR. No need to review yet

**Pull request closed without merging**